plutus-core/plutus-ir/src/PlutusIR/Parser.hs

```diff
@@ -29,109 +28,113 @@ import PlutusCore.Error (AsParserErrorBundle)
 
 conTerm :: Parametric
-conTerm _tm = inParens $ PIR.constant <$> wordPos "con" <*> constant
+conTerm _tm = withSpan $ \sp ->
+    inParens $ PIR.constant sp <$> (symbol "con" *> constant)
 
 iwrapTerm :: Parametric
-iwrapTerm tm = inParens $ PIR.iWrap <$> wordPos "iwrap" <*> pType <*> pType <*> tm
+iwrapTerm tm = withSpan $ \sp ->
+    inParens $ PIR.iWrap sp <$> (symbol "iwrap" *> pType) <*> pType <*> tm
 
 builtinTerm :: Parametric
-builtinTerm _tm = inParens $ PIR.builtin <$> wordPos "builtin" <*> builtinFunction
+builtinTerm _tm = withSpan $ \sp ->
+    inParens $ PIR.builtin sp <$> (symbol "builtin" *> builtinFunction)
 
 unwrapTerm :: Parametric
-unwrapTerm tm = inParens $ PIR.unwrap <$> wordPos "unwrap" <*> tm
+unwrapTerm tm = withSpan $ \sp ->
+    inParens $ PIR.unwrap sp <$> (symbol "unwrap" *> tm)
 
 errorTerm :: Parametric
-errorTerm _tm = inParens $ PIR.error <$> wordPos "error" <*> pType
+errorTerm _tm = withSpan $ \sp ->
+    inParens $ PIR.error sp <$> (symbol "error" *> pType)
 
-letTerm
-    :: Parser PTerm
-letTerm = Let <$> wordPos "let" <*> recursivity <*> NE.some (try binding) <*> pTerm
+letTerm :: Parser PTerm
+letTerm = withSpan $ \sp ->
+    inParens $ Let sp <$> (symbol "let" *> recursivity) <*> NE.some (try binding) <*> pTerm
 
 appTerm :: Parametric
-appTerm tm = do
-    pos <- getSourcePos
-    inBrackets $ PIR.mkIterApp <$> pure pos <*> tm <*> some tm
+appTerm tm = withSpan $ \sp ->
+    inBrackets $ PIR.mkIterApp sp <$> tm <*> some tm
 
 tyInstTerm :: Parametric
-tyInstTerm tm = do
-    pos <- getSourcePos
-    inBraces $ PIR.mkIterInst <$> pure pos <*> tm <*> some pType
-
-term' :: Parametric
-term' other = choice $ map try [
-    varTerm
-    , absTerm self
-    , lamTerm self
-    , conTerm self
-    , iwrapTerm self
-    , builtinTerm self
-    , unwrapTerm self
-    , errorTerm self
-    , inParens other
-    , tyInstTerm self
-    , appTerm self
-    ]
-    where self = term' other
+tyInstTerm tm = withSpan $ \sp ->
+    inBraces $ PIR.mkIterInst sp <$> tm <*> some pType
 
 pTerm :: Parser PTerm
-pTerm = whitespace >> term' letTerm
+pTerm = whitespace >> go
+  where
+    go = choice $ try <$>
+        [ varTerm
+        , letTerm
+        , absTerm go
+        , lamTerm go
+        , conTerm go
+        , iwrapTerm go
+        , builtinTerm go
+        , unwrapTerm go
+        , errorTerm go
+        , tyInstTerm go
+        , appTerm go
+        ]
 
 -- Note that PIR programs do not actually carry a version number
 -- we (optionally) parse it all the same so we can parse all PLC code
-program :: Parser (Program TyName Name PLC.DefaultUni PLC.DefaultFun SourcePos)
-program = whitespace >> do
-    prog <- inParens $ do
-        p <- wordPos "program"
-        option () $ void version
-        Program p <$> pTerm
+program :: Parser (Program TyName Name PLC.DefaultUni PLC.DefaultFun SrcSpan)
+program = do
+    whitespace
+    prog <- withSpan $ \sp ->
```

Comment:
For consistency, I wonder if we should pull this bit out as the program parser and then have a "top-level" parser that does the "leading whitespace and consumes all input" check.

Reply:
Sure.